src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl_ModelComputation.cs

```diff
@@ -0,0 +1,256 @@
+        /// Starts a new task to compute the data model.
+        /// </summary>
+        private void StartComputeDataModelTask()
+        {
+            // 'true' value is unused.  this just signals to the queue that we have work to do.
+            _computeDataModelQueue.AddWork(true);
+        }
+
+        /// <summary>
+        /// Creates the data model.
+        /// </summary>
+        private async ValueTask<DocumentSymbolModel?> ComputeDataModelAsync(ImmutableSegmentedList<bool> _, CancellationToken cancellationToken)
+        {
+            // Jump to the UI thread to get the currently active text view.
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var activeTextView = GetLastActiveIWpfTextView();
+            if (activeTextView is null)
+                return null;
+
+            var originalSnapshot = activeTextView.TextSnapshot;
+            var textBuffer = activeTextView.TextBuffer;
+
+            var filePath = GetFilePath();
+            if (filePath is null)
+                return null;
+
+            // Ensure we switch to the threadpool before calling ComputeUIModelAsync. It ensures
+            // that fetching and processing the document model is not done on the UI thread.
+            await TaskScheduler.Default;
+
+            var model = await ComputeDataModelAsync().ConfigureAwait(false);
+
+            if (model is not null)
+                StartUpdateDataModelTask();
+
+            return model;
+
+            async Task<DocumentSymbolModel?> ComputeDataModelAsync()
+            {
+                var response = await DocumentOutlineHelper.DocumentSymbolsRequestAsync(
+                    textBuffer, LanguageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
+
+                if (response is null)
+                    return null;
+
+                var responseBody = response.ToObject<DocumentSymbol[]>();
+                if (responseBody is null)
+                    return null;
+
+                var documentSymbols = DocumentOutlineHelper.GetNestedDocumentSymbols(responseBody);
+                var documentSymbolItems = DocumentOutlineHelper.GetDocumentSymbolItems(documentSymbols);
+                return new DocumentSymbolModel(documentSymbolItems, originalSnapshot);
+            }
+
+            string? GetFilePath()
+            {
+                ThreadingContext.ThrowIfNotOnUIThread();
+                if (EditorAdaptersFactoryService.GetBufferAdapter(textBuffer) is IPersistFileFormat persistFileFormat &&
+                    ErrorHandler.Succeeded(persistFileFormat.GetCurFile(out var filePath, out var _)))
+                {
+                    return filePath;
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Starts a new task to update the UI model.
+        /// </summary>
+        private void StartUpdateDataModelTask()
+        {
```

Comment:
Does the method need to be renamed? data model vs. UI model mean different things or no?